compat_test/compat_test.go

```diff
@@ -0,0 +1,1012 @@
+	}
+	finalEncodedString := hexutil.Encode(append(extraDataBytes[:IstanbulExtraVanity], reEncodedExtra...))
+
+	block["extraData"] = finalEncodedString
+
+	return nil
+}
+
+func (r *blockResults) verifyTransactions() error {
+	makeTransactionsComparable(r.celoTxs)
+	makeTransactionsComparable(r.opTxs)
+
+	// We also need to take into account yparity (only set on op transactions which can cause different representations for the v value)
+	// We overcome this by re-setting the v value of the celo txs to the op txs v value
+	for _, tx := range r.celoTxs {
+		// This sets the y value to be a big number where abs is nul rather than a zero length array if the numebr is zero.
+		// It doesn't change the number but it does change the representation.
+		types.SetYNullStyleBigIfZero(tx)
+	}
+	err := EqualObjects(r.celoTxs, r.opTxs)
+	if err != nil {
+		return err
+	}
+	// filter raw celo and op txs
+	for i := range r.celoRawTxs {
+		filterCeloTx(r.celoRawTxs[i])
+	}
+	for i := range r.opRawTxs {
+		filterOpTx(r.opRawTxs[i])
+	}
+	err = EqualObjects(r.celoRawTxs, r.opRawTxs)
+	if err != nil {
+		return err
+	}
+
+	// cross check txs, unfortunately we can't easily do a direct comparison here so we compare number of txs and their
+	// hashes.
+	err = EqualObjects(len(r.celoTxs), len(r.celoRawTxs))
+	if err != nil {
+		return err
+	}
+	for i := range r.celoTxs {
+		err = EqualObjects(r.celoTxs[i].Hash().String(), r.celoRawTxs[i]["hash"].(string))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Cross check the individually retrieved transactions with the transactions in the block
+	return EqualObjects(r.celoTxs, []*types.Transaction(r.opBlockByNumber.Transactions()))
+}
+
+func (r *blockResults) verifyReceipts() error {
+	err := EqualObjects(r.celoReceipts, r.opReceipts)
+	if err != nil {
+		return err
+	}
+
+	// filter the raw op receipts
+	for i := range r.opReceipts {
+		filterOpReceipt(r.opRawReceipts[i])
+	}
+	err = EqualObjects(len(r.celoRawReceipts), len(r.opRawReceipts))
+	if err != nil {
+		return err
+	}
+	for i := range r.celoRawReceipts {
+		err = EqualObjects(r.celoRawReceipts[i], r.opRawReceipts[i])
+		if err != nil {
+			if r.celoRawReceipts[i]["effectiveGasPrice"] != nil && r.opRawReceipts[i]["effectiveGasPrice"] == nil {
+				fmt.Printf("dangling state at block %d\n", r.blockNumber-1)
+			} else {
+				return err
+			}
+		}
+	}
+	// Cross check receipts, here we decode the raw receipts into receipt objects and compare them, since the raw
+	// receipts are enriched with more fields than the receipt objects.
+	var celoRawConverted []*types.Receipt
+	jsonConvert(r.celoRawReceipts, &celoRawConverted)
+	err = EqualObjects(celoRawConverted, r.celoReceipts)
+	if err != nil {
+		spew.Dump("celorawreceipts", r.celoRawBlockReceipts)
+		return err
+	}
+
+	for _, opReceipt := range r.opRawReceipts {
+		if opReceipt["cumulativeGasUsed"] == 0 {
+			return fmt.Errorf("receipt with hash %q has CumulativeGasUsed == 0", opReceipt["transactionHash"])
+		}
+	}
```

Comment:
Yep, at one point I thought I discovered a receipt with a 0 cumulative gas used, but probably I was mistaken because on re-runs I never found one, so this can be removed.